pkg/operator/controller/ingress/status.go

```diff
@@ -171,35 +182,101 @@ func computeDeploymentPodsScheduledCondition(deployment *appsv1.Deployment, pods
 }
 
 // computeIngressAvailableCondition computes the ingress controller's current Available status state
-// by inspecting the Available condition of deployment. The ingresscontroller is only available if
-// the deployment is also available.
-func computeIngressAvailableCondition(deployment *appsv1.Deployment) operatorv1.OperatorCondition {
-	for _, cond := range deployment.Status.Conditions {
-		if cond.Type != appsv1.DeploymentAvailable {
+// by inspecting the following:
+// 1) the Available condition of Deployment,
+// 2) the DNSReady condition of the IngressController, and
+// 3) the LoadBalancerReady condition of the IngressController.
+// The ingresscontroller is judged Available only if all 3 conditions are true
+func computeIngressAvailableCondition(conditions []operatorv1.OperatorCondition) operatorv1.OperatorCondition {
+	expected := []expectedCondition{
+		{
+			condition: IngressControllerDeploymentAvailableConditionType,
+			status:    operatorv1.ConditionTrue,
+		},
+		{
+			condition: operatorv1.DNSReadyIngressConditionType,
+			status:    operatorv1.ConditionTrue,
+			ifConditionsTrue: []string{
+				operatorv1.LoadBalancerManagedIngressConditionType,
+				operatorv1.LoadBalancerReadyIngressConditionType,
+				operatorv1.DNSManagedIngressConditionType,
+			},
+		},
+		{
+			condition:        operatorv1.LoadBalancerReadyIngressConditionType,
+			status:           operatorv1.ConditionTrue,
+			ifConditionsTrue: []string{operatorv1.LoadBalancerManagedIngressConditionType},
+		},
+	}
+
+	// Cover the rare case of no conditions
+	if len(conditions) == 0 {
+		return operatorv1.OperatorCondition{Type: operatorv1.OperatorStatusTypeAvailable, Status: operatorv1.ConditionFalse}
+	}
+	_, unavailableConditions, _ := checkConditions(expected, conditions)
+	if len(unavailableConditions) != 0 {
+		degraded := formatConditions(unavailableConditions)
+		return operatorv1.OperatorCondition{
+			Type:    operatorv1.IngressControllerAvailableConditionType,
+			Status:  operatorv1.ConditionFalse,
+			Reason:  "IngressControllerUnavailable",
+			Message: "One or more status conditions indicate unavailable: " + degraded,
+		}
+	}
+	return operatorv1.OperatorCondition{
+		Type:   operatorv1.IngressControllerAvailableConditionType,
+		Status: operatorv1.ConditionTrue,
+	}
+}
+
+// checkConditions compares expected operator conditions to existing operator
+// conditions and returns a list of graceConditions, degradedconditions, and a
+// requeueing wait time.
+func checkConditions(expecteds []expectedCondition, conditions []operatorv1.OperatorCondition) ([]*operatorv1.OperatorCondition, []*operatorv1.OperatorCondition, time.Duration) {
```

Comment:
thoughts on naming this parameter `expectedConds` instead of `expecteds` (or something along those lines)?